lib/vast/Tower/Link.cpp

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2024-present, Trail of Bits, Inc.
+
+#include "vast/Tower/Link.hpp"
+
+#include <ranges>
+
+namespace vast::tw {
+
+    op_mapping reverse_mapping(const op_mapping &from) {
+        op_mapping out;
+        for (const auto &[root, ops] : from) {
+            for (auto op : ops) {
+                out[op].push_back(root);
+            }
+        }
+        return out;
+    }
+
+    void dbg(const op_mapping &mapping, auto &outs) {
+        outs << "Mapping:\n";
+
+        auto flag = mlir::OpPrintingFlags().skipRegions();
+        auto render_op = [&](operation op) -> decltype(outs) & {
+            outs << op << ": ";
+            op->print(outs, flag);
+            if (!op->getRegions().empty())
+                outs << " ... regions ...";
+            return outs;
+        };
+
+        for (const auto &[from, to] : mapping) {
+            outs << "== ";
+            render_op(from) << "\n";
+            for (auto t : to) {
+                outs << "   -> ";
+                render_op(t) << "\n";
+            }
+        }
+    }
+
+    using loc_to_op_t = llvm::DenseMap< loc_t, operations >;
+
+    void dbg(const loc_to_op_t &mapping, auto &outs) {
+        outs << "loc_to_op\n";
+        for (const auto &[loc, to] : mapping) {
+            outs << ".." << loc << "\n";
+            for (auto t : to) {
+                outs << "   -> " << *t << "\n";
+            }
+        }
+    }
+
+    loc_to_op_t gather_map(location_info_t &li, operation op) {
+        loc_to_op_t out;
+        auto collect = [&](operation op) { out[li.self(op)].push_back(op); };
+        op->walk(collect);
+        return out;
+    }
+
+    std::vector< op_mapping > mk_link_mappings(const auto &links) {
```

Comment:
```
auto mk_link_mappings(const std::ranges::range &links) {
    return links | std::views::transform([](const auto &l) {
        return l->parents_to_children();
    }) | std::ranges::to<std::vector>();
}
``` 

and for cleaner algorithms we might want to provide: 

```
static inline auto parents_to_children = [] (const auto &link) {
    return link->parents_to_children();
};
```

the result is then:
```
auto mk_link_mappings(const std::ranges::range &links) {
    return links | vws::transform(parents_to_children) | rns::to< std::vector >();
}
``` 

where 

```
    namespace rns = std::ranges;
    namespace vws = std::ranges::views;
```

I use these in new symbol tables (we can discuss better abbreviations, 
but we should standardize one across entire project.